lib/llvmopencl/linker.cpp

```diff
@@ -262,46 +264,65 @@ static void fixCallingConv(llvm::Module *Mod, std::string &Log) {
   }
 }
 
-// Find all functions in the calltree of F, append their
-// name to function name set.
-static inline void
+using SmallFunctionSet = llvm::SmallSet<llvm::Function *, 8>;
+
+// Find all functions in the calltree of F, including declarations.
+// Returns the list of Functions called in CalledFuncList,
+// in depth-first order (to make cloning simpler);
+// returns -1 on error (recursion), 0 on success
+static int
 find_called_functions(llvm::Function *F,
-                      llvm::StringSet<> &FNameSet)
-{
+                      llvm::SmallVector<llvm::Function *> &CalledFuncList,
+                      SmallFunctionSet &CallStack) {
   if (F->isDeclaration()) {
-    DB_PRINT("it's a declaration.\n");
-    return;
+    DB_PRINT("it's a declaration, return\n");
+    return 0;
   }
 
-  llvm::Function::iterator FI;
-  for (FI = F->begin(); FI != F->end(); FI++) {
-    llvm::BasicBlock::iterator BI;
-    for (BI = FI->begin(); BI != FI->end(); BI++) {
-      CallInst *CI = dyn_cast<CallInst>(BI);
-      if (CI == NULL)
-        continue;
-      llvm::Function *Callee = CI->getCalledFunction();
-      // this happens with e.g. inline asm calls
-      if (Callee == NULL) {
-        DB_PRINT("search: %s callee NULL\n", F->getName().str().c_str());
-        continue;
-      }
-      if (!Callee->hasName()) {
-        Callee->setName("__anonymous_function");
-      }
-      const char* Name = Callee->getName().data();
-      DB_PRINT("search: %s calls %s\n",
-               F->getName().data(), Name);
-      if (FNameSet.count(Callee->getName()) > 0)
-        continue;
-      else {
-        DB_PRINT("inserting %s\n", Name);
-        FNameSet.insert(Callee->getName());
-        DB_PRINT("search: recursing into %s\n", Name);
-        find_called_functions(Callee, FNameSet);
-      }
+  CallStack.insert(F);
+
+  assert(F->hasName());
+  std::string FName = F->getName().str();
+
+  for (auto &I : instructions(F)) {
```

Comment:
Good to know there's this type of an iterator-generator.

Reply:
Yep this pattern (iterating over all instructions of a function) is so common they've made an iterator for it